### DISCUSSION TITLE
Improved Hoarding & Data Storage System

### DISCUSSION THREAD

**Discussion by adamhannigan** [sunflower-land/sunflower-land#2527] [Ideas]
The current hoarding system has a variety of challenges and adds a lot of negative sentiment and friction while playing Sunflower Land. The purpose of this discussion is to explore an improved mechanic.

## On Chain Data

We are a Web3 game and deeply believe in the positive benefits that arise from on-chain data. Our goal in Sunflower Land is to ensure that the appropriate data is stored on the Blockchain. 

The purpose of the hoarding limits are too ensure that our on-chain data, closely reflects what is happening in the game. The benefits of this include:

- Increased Security
- Transparent Data - on chain leaderboards + empowering open source developers to build tools + products on top of our game.
- Players can use there resources in Web3 environments (Goblin Retreat)

## Current Problems with Hoarding Limits & Storing Data on Chain

1. High Data Storage - Storing all of your game progress on chain requires a lot of gas. As our game evolves and becomes more complex, the amount of data we need to store will also increase. Unless addressed, this means that players will need to pay more gas in their transactions
2. Rising Gas Prices - Polygon network prices are constantly increasing and at times has excessive peak periods where transaction prices are unaffordable
3. Gaming friction - No one wants to be interrupted while in the middle of farming. The current hoarding system has an obtrusive modal that blocks game play. It is near impossible to know when this modal will appear and can disrupt your planning.
4. Confusing User Experience - For new players, it is unclear what the purpose of the Hoarding system and how it benefits them. Visiting islands such as Goblin Retreat also adds a lot of confusion - "Why did my items disappear?"

# Proposed Solution

The proposed solution aims to address a range of the problems above.

## Silo

We will introduce the concept of a Silo building in Sunflower Land. This building will display your resources that are on chain and also the limits of your current in progress resources. It will let players know which exact resource are reaching their max limits so players can plan accordingly

The narrative will revolve around players hitting max limits in their inventory and being urged to store their resources into the Silo before continuing. The Silo is a metaphor for your secure Blockchain data.

When visiting Goblin Retreat, we will use icons for the Silo instead of your normal Inventory icon. The purpose is to promote the clear distinction that only Silo items are available at the Retreat.

## Opt in Data Storage

We will give players the option to choose which items they want to store in their Silo. This means you only need to store items that are nearing their hoarder limits or ones that you want to store on chain for use at Goblin Retreat.

This significantly reduces the gas price and also means the most important items are being stored on chain.

A downside is that less data is stored on chain. However, we believe the most critical data that is used in popular on-chain leaderboards & used for trading will continue being stored on chain. It is only insignificant items (such as buildings) that will not reflect the real-time nature of the game.

## Conclusion

Thanks for taking the time to read and participate in this discussion 🤗 

**Note:** _This solution excludes gas optimisation + smart contract improvements. These will constantly be added and improved in our game. Other technologies such as Supernets and ZkEvm are currently being investigated but not yet planned for implementation. We also exclude fixing the appropriate hoarder limits per item, this will be done separately to this discussion._

**Comment by TheDcol** [comment]
> We will give players the option to choose which items they want to store in their Silo. This means you only need to store items that are nearing their hoarder limits or ones that you want to store on chain for use at Goblin Retreat.

Okay, I can get behind this, especially if storing 1 type of resource costs VERY little matic. If storing a resource costs .01 matic then hoard limits could be left as they are (although I'd like to see them increase regardless, especially after seeing the 20 BB buy button in PRs)

> A downside is that less data is stored on chain

Honestly, if it makes for a better gaming experience, I personally would be willing to sacrifice that a little bit.

In regards to hoard limits, I have been seeing in Discord people complaining about the usual resources that 'block' them from progressing and force them to sync. With upcoming boosts such as farm items, sprinklers, tractors, hay bales, skills new nfts etc. we are bound to be hitting the hoard limit more and more as the game progresses.

Below is an image showing proposed increases to hoard limits which should cover upcoming boosts, it is a draft version and does not include eggplant nfts.

![image](https://user-images.githubusercontent.com/108110065/235822926-3efa9a9f-212b-49bb-aa55-c2a797979b31.png)

Just wanted to share that randomly. Thanks for looking into the hoarding/storage system as it has become a hot topic for many in Discord.

**Reply by adamhannigan** [reply]
Thanks mate @TheDcol We will include these in today's update 👍

**Reply by vitt0** [reply]
Something that would be good to also increase is the Dawn Breaker Tickects, the limit is at 350 I think it's safe to raise it to at least 500

**Comment by SaYaNiTo** [comment]
Guys. I bought 5 BB, i got 2 left and i m hoardering on cabbages since the middle of my current ticket. Going to 1000 to 1500 on cabbage with 20 tickets , really ?
It is that hard to adapt hoardering limits to be dynamic ?

**Reply by TheDcol** [reply]
I based those numbers from 5 restocks, not 20.
The Silo building above allowing you to store 1 type of resource for a very low fee (hopefully) would lessen the hoarding issue anyway

> It is that hard to adapt hoardering limits to be dynamic ?

Feel free to set it up mate if it's not hard! :D

**Reply by SaYaNiTo** [reply]
> Feel free to set it up mate if it's not hard! :D

Not working for sunflower, plus i m not a developper. Just have some knowledge that makes me thinks that it should be doable. Just asking if it is or not.

> I based those numbers from 5 restocks, not 20.

For 5 restocks, it seems close to be ok. I m not sure about fruits. If someone does full one fruit with immortal pear : 12*4 = 48 * 5 = 240

**Reply by TheDcol** [reply]
> For 5 restocks, it seems close to be ok. I m not sure about fruits. If someone does full one fruit with immortal pear : 12*4 = 48 * 5 = 240

Regarding fruit: 10 seeds per restock. 10*5 BBs = 50 seeds in total. 50*4 = 200

It's 250 with immortal pear, which does go over the 200 hoard limit, still a lot better than the current 100 limit

> Just asking if it is or not.

Fair enough, I'd be curious too. What would you tie down the dynamic part to? 
When I proposed costs/funcction for the hen house, i initially added an increased egg hoarding limit. So the hoard limit is increased depending on how many hen houses you have. Similarly for wharehouse and toolshed

**Reply by SaYaNiTo** [reply]
> Regarding fruit: 10 seeds per restock. 10_5 BBs = 50 seeds in total. 50_4 = 200

it s 12 seeds per stock with warehouse :)

**Reply by SaYaNiTo** [reply]
![image](https://user-images.githubusercontent.com/124063442/235853303-3e43b3d2-6568-4445-b9ed-afd8c054ba1b.png)
just discovered that ... Still got 2 tickets to use ... i ll hit it by the end of the day probably

**Reply by TheDcol** [reply]
Yup, that increased 
![Screenshot_20230504-203400](https://user-images.githubusercontent.com/108110065/236180359-50c1a06c-6922-4557-bb8c-d01d86c730ea.png)

**Reply by adamhannigan** [reply]
Hey @SaYaNiTo Can you please provide more details around your current farming operations.

What outcome are you aiming for and why do you have so many cabbages? These hoarding limits are designed around burns and sinks in the game, namely using these crops in recipes such as the Cabbage Cake, Saurrekrat etc.

**Comment by brian-man** [comment]
Are there any plans for the `Daily SFL Limit`?  It's kind of surprising and terrible UX right now.

**Reply by adamhannigan** [reply]
This is mostly a security feature to safe guard any issues that could arise.

For most farming operations this doesn't affect people - even on the off chance of finding 2 Pirate Bounties in a single day.

We do see people facing this issue who have hoarded lots of crops and are looking to sell in one go. For instance, 10,000 Radishes at once. This is an interested case that we need to factor in a solution for.

**Reply by brian-man** [reply]
> We do see people facing this issue who have hoarded lots of crops and are looking to sell in one go. For instance, 10,000 Radishes at once. This is an interested case that we need to factor in a solution for.

Also it's going to be a "poop emoji" show when half the player base gets locked out of the game repeatedly as the halving approaches and they want to liquidate their crops before they drop in value by 50%.

**Comment by offerT** [comment]
Since you can choose which things to upload, I believe no one will choose to upload a seed. Seeds do not need to be stored on the chain in the future

**Reply by adamhannigan** [reply]
Agreed 💯

**Reply by brian-man** [reply]
> Since you can choose which things to upload, I believe no one will choose to upload a seed. Seeds do not need to be stored on the chain in the future

Yeah, I mentioned this previously. :)
<https://github.com/sunflower-land/sunflower-land/discussions/2466>

**Reply by eliasSFL** [reply]
Going forward, likely the things that generally players will be sync are:
- Resources
- Crops
- Special event tickets
- Occasional SFTs for withdrawal

The rest can remain on the server side.
Any plans to remove hoarding limit for all non-essential items? _(Of course the limit for holding a certain amount of seeds should still remain)_

**Comment by TheDcol** [comment]
Silo is a great idea and should be implemented asap 💯☕

**Reply by SaYaNiTo** [reply]
WEN silo ^^ 
Won t be perfect but so useful, can t wait to see how much ressourcesit gonna cost

**Reply by eliasSFL** [reply]
@SaYaNiTo since only resources are the concern of players moving forward only letting us select whatever we want to sync will reduce gas fees by alot. I'm really looking forward to this Silo update.

**Comment by brian-man** [comment]
General feedback:
Does the current design (and/or security concern) require storing a resource on chain to reset hoarder for that resource?

As an example, I want to keep farming sunflowers, potatoes, etc. for future mints even though I don't really need them to be on-chain for months (years?) until that mint is introduced.  But hoarder forces me to sync them anyway.

**Reply by eliasSFL** [reply]
@brian-man Hoard limits will still be a thing but because you can choose to sync just the item on chain when u hit the hoard limit (rather than storing everything), hence reducing the amount of data being stored on the blockchain will make gas fees a lot cheaper

**Reply by brian-man** [reply]
> @brian-man Hoard limits will still be a thing but because you can choose to sync just the item on chain when u hit the hoard limit (rather than storing everything), hence reducing the amount of data being stored on the blockchain will make gas fees a lot cheaper

I'm aware that's the intended design so far, but that doesn't answer my question.

From a technical perspective, resetting the hoarder limits should only require the smallest (gas) possible transaction that ETHVM blockchains support.  Storing changes to SLI items is not necessary to reset the limits, from a technical perspective.

If the goal is truly to make playing the game "gas inexpensive" then a "reset hoarder limits" transaction with zero SLI updates should be provided to players.

**Reply by eliasSFL** [reply]
> > @brian-man Hoard limits will still be a thing but because you can choose to sync just the item on chain when u hit the hoard limit (rather than storing everything), hence reducing the amount of data being stored on the blockchain will make gas fees a lot cheaper
> 
> I'm aware that's the intended design so far, but that doesn't answer my question.
> 
> From a technical perspective, resetting the hoarder limits should only require the smallest (gas) possible transaction that ETHVM blockchains support.  Storing changes to SLI items is not necessary to reset the limits, from a technical perspective.

Hoard limits is the difference between what you have on chain and what's on the server. Storing them on chain will match the items on the server to on chain, hence resetting the limit.
Say you have 2000 sunflowers on chain and you have 11000 sunflowers on the server, hence hitting the limit of 9000. When you store your sunflowers on chain now you have 11000 sunflowers on chain and 11000 sunflowers in the server, hence resetting the hoard limit for sunflowers.
And cuz you're only syncing the item that you want to reset the hoard limit for, reducing the gas fees needed to sync on chain

**Reply by adamhannigan** [reply]
Our desired gameplay is that enough sinks exist in game that players keep burning their resources. Hoarding resources for a future use case indicates a deeper problem of a lack of utility for crops in the game.

If players are trying to hoard 10+ resources for the use case you mentioned, they will need to store these on chain.

**Reply by brian-man** [reply]
> Hoarding resources for a future use case indicates a deeper problem of a lack of utility for crops in the game.

I disagree.  The kid mints from last season were 2000 Cabbage.  That motivates people to hoard so that they are in a better position for the next one that will be 5000, or 10000, or 20000.

Further, "convert to SFT, withdraw SFT, sell SFT" is the "how to pay for gas" model that is promoted.  This puts even more weight into "save up enough crops to mint the next valuable SFTs" vs. "use or withdraw your crops at 30% tax".

**Reply by FarmerKuenster** [reply]
another thing to add to this, players who collected a ton of resources (for the reason @brian-man listed above) is something ive done. then when i decide i dont want to hoard any more ill sell some crops. I now get penalized with a hoarding sfl message and cant sell more in game until the next day. Now the game makes it feel kind of like a damned if you do, damned if you don't.  this is just my personal opinion.

**Comment by JCtapuk** [comment]
I think storage containers are better.
If you have 1000 sunflowers and want to store them on the blockchain, then you need to create a container.

Container 1 pc. And requires wood 5 pieces to create.
The container item is not stored in the blockchain, but you can send certain goods to the bunker, for example
1000 sunflowers and 1 container. This will reduce the amount and severely burn wood when the player wants to pull items out of the bunker with a container.
